HARK/ConsumptionSaving/ConsLabeledModel.py

```diff
@@ -0,0 +1,663 @@
+from dataclasses import dataclass
+from types import SimpleNamespace
+
+import numpy as np
+import xarray as xr
+from HARK.ConsumptionSaving.ConsIndShockModel import (
+    ConsIndShockSetup,
+    IndShockConsumerType,
+    init_perfect_foresight,
+)
+from HARK.ConsumptionSaving.ConsRiskyAssetModel import (
+    RiskyAssetConsumerType,
+    init_risky_asset,
+)
+from HARK.core import MetricObject, make_one_period_oo_solver
+from HARK.distribution import DiscreteDistributionLabeled
+from HARK.utilities import CRRAutility, CRRAutility_inv, CRRAutilityP, CRRAutilityP_inv
+
+
+class ValueFuncCRRALabeled(MetricObject):
+    def __init__(self, dataset: xr.Dataset, CRRA: float):
+
+        self.dataset = dataset
+        self.CRRA = CRRA
+
+    def __call__(self, state):
+        """
+        Interpolate inverse falue function then invert to get value function at given state.
+        """
+
+        if isinstance(state, xr.Dataset):
+            state_dict = state.coords
+        else:
+            state_dict = {}
+            for coords in self.dataset.coords.keys():
+                state_dict[coords] = state[coords]
+
+        result = CRRAutility(
+            self.dataset["v_inv"].interp(
```

Comment:
I am confused about what mathematical object "v_inv" refers to.

It does not appear to be the mathematical inverse of the value function, which would be my first guess.

Reply:
In the last period, $v(m) = u(c(m)) = u(m)$, so we define the "inverse" as $v^{-1}(m) = u^{-1}(v(m)) \equiv m$. 

In this case $v$ is curved but $v^{-1}$ is linear, so if we want to approximate $v$ off the grid we get better results linearly approximating $v^{-1}$ then recurving, than linearly interpolating $v$, especially at low values of $m$. 

We can do the same at earlier periods, where $v^{-1}_t(m) = u^{-1}(v_t(m))$. In For $t \ne T$, $v^{-1}_t$ is not linear, but it's much less curved than $v_t$ (does not go to neg infty) so a linear approximation approach is still better.